storage/filestore/dir.go

```diff
@@ -238,6 +348,136 @@ func (dir *Dir) GarbageCollect(ctx context.Context) (err error) {
+				// don't need to pass on this error
+				continue
+			}
+			ids = append(ids, namespace)
+		}
+	}
+}
+
+// ForAllKeysInNamespace executes doForEach for each locally stored blob, stored with storage
+// format V1 or greater, in the given namespace. If doForEach returns a non-nil error,
+// ForAllKeysInNamespace will stop iterating and return the error immediately.
+func (dir *Dir) ForAllKeysInNamespace(ctx context.Context, namespace []byte, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	namespaceDir := pathEncoding.EncodeToString(namespace)
+	nsDir := filepath.Join(dir.blobsdir(), namespaceDir)
+	openDir, err := os.Open(nsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// job accomplished: there are no blobs in this namespace!
+			return nil
+		}
+		return err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		// check for context done both before and after our readdir() call
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		subdirNames, err := openDir.Readdirnames(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if os.IsNotExist(err) || len(subdirNames) == 0 {
+			return nil
+		}
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		for _, keyPrefix := range subdirNames {
+			if len(keyPrefix) != 2 {
+				// just an invalid subdir; could be garbage of many kinds. probably
+				// don't need to pass on this error
+				continue
+			}
+			err := dir.forAllKeysInNamespaceWithPrefix(ctx, namespace, nsDir, keyPrefix, doForEach)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (dir *Dir) forAllKeysInNamespaceWithPrefix(ctx context.Context, namespace []byte, nsDir, keyPrefix string, doForEach func(storage.StoredBlobAccess) error) (err error) {
+	keyDir := filepath.Join(nsDir, keyPrefix)
+	openDir, err := os.Open(keyDir)
+	if err != nil {
+		return err
+	}
+	defer func() { err = errs.Combine(err, openDir.Close()) }()
+	for {
+		// check for context done both before and after our readdir() call
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		keyInfos, err := openDir.Readdir(nameBatchSize)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if os.IsNotExist(err) || len(keyInfos) == 0 {
+			return nil
+		}
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		for _, keyInfo := range keyInfos {
+			if keyInfo.Mode().IsDir() {
+				continue
+			}
+			blobFileName := keyInfo.Name()
+			encodedKey := keyPrefix + blobFileName
+			formatVer := storage.FormatV0
+			if strings.HasSuffix(blobFileName, v1PieceFileSuffix) {
+				formatVer = storage.FormatV1
+				encodedKey = encodedKey[0 : len(encodedKey)-len(v1PieceFileSuffix)]
+			}
+			key, err := pathEncoding.DecodeString(encodedKey)
+			if err != nil {
+				continue
+			}
+			ref := storage.BlobRef{
+				Namespace: namespace,
+				Key:       key,
+			}
+			fullPath := filepath.Join(keyDir, blobFileName)
+			err = doForEach(newStoredBlobAccess(ref, fullPath, keyInfo, formatVer))
```

Comment:
I think we should pass `ctx` to `doForEach` as well.

Reply:
I thought about that. But if the ctx passed in to this function is the right one to apply to the callback, it's very easy for the caller to provide the same ctx to the callback via closure. So it seems to cover more cases if we don't pass along ctx. But I don't care too much.

Reply:
The difference is that the filestore could potentially change the context at some intermediate step, in which case the parent context might not be the best thing. Hence the reason for passing it down, rather than relying on the call location context.

Reply:
How about if I document specifically that the context parameter is intended only for the purpose of being able to cancel the directory traversal operation? Would that help instead?

Reply:
Yeah, we can add a comment.